artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/SimpleAddressManager.java

```diff
@@ -100,6 +104,11 @@ public boolean addBinding(final Binding binding) throws Exception {
       if (nameMap.putIfAbsent(binding.getUniqueName(), bindingAddressPair) != null) {
          throw ActiveMQMessageBundle.BUNDLE.bindingAlreadyExists(binding);
       }
+      directBindingMap.compute(binding.getAddress(), (key, value) -> {
+         Collection<Binding> bindingList = value == null ? new ArrayList<>() : value;
+         bindingList.add(binding);
```

Comment:
Maybe I'm being naive... but as far as I remember.. the ArrayList itself is not supposed to be used from multiple threads.

The compute will do a replay, but you could still call the compute from multiple threads.

I feel like you should have:

```
    synchronized(bindingList) {
        bindingsList.add(binding);
    }
```


Or another data structure to be used here.


similarly on the remove the queue, it should use a synchronized or use a concurrent equivalent.

Reply:
say you call this from different threads... 


perhaps we should add a test having mulitple threads on this call here, and make sure the outcome is as expected.

Reply:
The `directBindingMap` is a `ConcurrentHashMap` so it will guarantee thread safety on the `compute` by locking the node. https://stackoverflow.com/a/59601136

Also, this should be tested already via `WildcardAddressManagerUnitTest#testConcurrentCalls`.

Reply:
Yeah.. I did some testing.. I thought the compute would do a loop retry only (I read the wrong code).


I added some testing.. and I changed your code with adding sleeps and everything worked.. I could see the wait to lock threads.


If you could rebase against main now that I added a test to validate this?


also, you can verify check style by doing this:

mvn -Pdev install -DskipTests=true -T2